File: examples/php/pwa/template.php
<?php
/**
 * Template wrapper for PHP example
 * Uses NeutralIpcTemplate to render templates via IPC.
 * See: https://github.com/FranBarInstance/neutralts-docs
 */

require_once __DIR__ . '/constants.php';
require_once __DIR__ . '/../ipc-client/NeutralIpcTemplate.php';

class Template
{
    protected $schema;

    public function __construct($schema)
    {
        $this->schema = $schema;
    }

    public function render()
    {
        $json = json_encode($this->schema);
        $template = new NeutralIpcTemplate(TEMPLATE_ROUTER, $json);
        $contents = $template->render();
        $status_code = intval($template->get_status_code());
        $status_text = $template->get_status_text();
        $status_param = $template->get_status_param();

        // Redirects from template
        if (in_array($status_code, [301, 302, 307, 308], true)) {
            header('Location: ' . $status_param, true, $status_code);
            exit;
        }

        // HTTP errors from template
        if ($status_code >= 400) {
            $error = [
                'data' => [
                    'CONTEXT' => ['ROUTE' => 'error'],
                    'error' => [
                        'code' => $status_code,
                        'text' => $status_text,
                        'param' => $status_param
                    ]
                ]
            ];
            $template->set_path(TEMPLATE_ERROR);
            $template->merge_schema(json_encode($error));
            $contents = $template->render();
        }

        http_response_code($status_code ?: 200);
        header('Content-Type: text/html');
        echo $contents;
        return null;
    }
}

File: examples/go/pwa/template.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"

	"neutral_ipc_template"
)

type Template struct {
	schema map[string]interface{}
}

func NewTemplate(schema map[string]interface{}) *Template {
	return &Template{schema: schema}
}

func (t *Template) Render(w http.ResponseWriter, r *http.Request) {
	jsonSchema, err := json.Marshal(t.schema)
	if err != nil {
		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
		return
	}

	nt := neutral_ipc_template.NewNeutralIpcTemplate(TEMPLATE_ROUTER, string(jsonSchema))
	contents := nt.Render()

	statusCode := nt.GetStatusCode()
	if statusCode != nil {
		codeStr := fmt.Sprintf("%v", statusCode)
		code, err := strconv.Atoi(codeStr)
		if err == nil {
			if code >= 300 && code < 400 {
				param := nt.GetStatusParam()
				paramStr := ""
				if p, ok := param.(string); ok {
					paramStr = p
				} else {
					paramStr = fmt.Sprintf("%v", param)
				}
				http.Redirect(w, r, paramStr, code)
				return
			}
			if code >= 400 {
				statusText := nt.GetStatusText()
				statusParam := nt.GetStatusParam()
				errorSchema := map[string]interface{}{
					"data": map[string]interface{}{
						"CONTEXT": map[string]interface{}{
							"ROUTE": "error",
						},
						"error": map[string]interface{}{
							"code":  float64(code),
							"text":  fmt.Sprintf("%v", statusText),
							"param": fmt.Sprintf("%v", statusParam),
						},
					},
				}
				jsonError, _ := json.Marshal(errorSchema)
				nt.SetPath(TEMPLATE_ERROR)
				nt.MergeSchema(string(jsonError))
				contents = nt.Render()
				w.WriteHeader(code)
			}
		}
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.Write([]byte(contents))
}
